pdudaemon/drivers/conrad197720.py

```diff
@@ -0,0 +1,325 @@
+        :param self: The object itself
+        """
+        byte = 0
+        if not self.com.is_open:
+            self.__openConnection(self.portname)
+        self.com.write(byte.to_bytes(1, byteorder='big'))
+        log.debug("sent single byte")
+
+    def __rxFrame(self, num_frames):
+        """
+        Private function to receive frame(s)
+        :param self: The object itself
+        :param num_frames: The number of frames to wait for (until timeout SERIAL_TIMEOUT occurs)
+        :return: The received data
+        :rtype: array
+        """
+        rx_data = []
+        if not self.com.is_open:
+            self.__openConnection(self.portname)
+        for i in range(0, (num_frames * 4)):
+            recv = self.com.read(1)
+            if len(recv) == 0:
+                break
+            recv = int.from_bytes(recv, 'little')
+            rx_data.append(recv)
+        log.debug(f"recv: {rx_data}")
+        return rx_data
+
+    def __checkFrameChecksum(self, data):
+        """
+        Private function to check the XOR checksum of the received frame
+        :param self: The object itself
+        :param data: The array holding the received data bytes, lenght must be a multiple of 4
+        :return: True, if checksum(s) match
+        :rtype: boolean
+        """
+        recv = ""
+        if len(data) % 4 != 0 or len(data) == 0:
+            return False
+        for i in range(0, (len(data) >> 2)):
+            # check if checksum matches
+            if (data[(i >> 2) + 0] ^ data[(i >> 2) + 1] ^ data[(i >> 2) + 2]) != data[(i >> 2) + 3]:
+                print(f"checksum of frame {i} does not match")
+                return False
+        return True
+
+    def __sendCommand(self, cmd_byte, addr_byte, data_byte):
+        """
+        Private function to send a command to the card(s) / retry / parse the response
+        :param self: The object itself
+        :param cmd_byte: The command to the card(s)
+        :param addr_byte: The card address, used when more than one relay card is connected to the same serial port
+        :param data_byte: The data byte to be transmitted
+        :return: the response data byte
+        :rtype: int
+        :raises RuntimeError: all retries failed, communication error
+        """
+        for i in range(0, RETRY_COUNT):
+            self.__txFrame([cmd_byte, addr_byte, data_byte])
+            num_frames = 1
+            # for init we might receive n+1 frames in case there n cards cascaded
+            if cmd_byte == CMD_INIT:
+                num_frames = MAX_CARDS + 1
+            recv_data = self.__rxFrame(num_frames)
+            # check if received data is valid
+            if len(recv_data) > 0 and self.__checkFrameChecksum(recv_data):
+                # retry, if there is no valid reply on init command
+                if cmd_byte == CMD_INIT:
+                    if (self.__checkNumCards(recv_data) <= 0):
+                        log.debug("__sendCommand retry reason 1")
```

Comment:
Would be good to describe the reason rather then just "reason 1"  for slightly more meaningful debug message messages

Reply:
good point, will be changed to log.debug("__sendCommand retry reason: no valid reply on init command")